File: Secuencial/Lector.py
from MacroBloque import MacroBloque
from PIL import Image

class Lector:

    #Constructor de la clase lector, recibe dos imagenes.
    def __init__(self, frame1, frame2):
        self.frame1 = frame1
        self.frame2 = frame2

    #Crea una matriz de pixeles ordenados en coordenadas x, y a partir de un arreglo de pixeles
    def llenarMatrizPixel(self, im):
        matrizPixeles = []
        for x in range(0,im.width):
            matrizPixeles.append([])
            for y in range(0,im.height):
                matrizPixeles[x].append(im.getpixel((x,y)))
        return matrizPixeles

    #Crea un arreglo de macrobloques para una imagen a partir de una matriz de pixeles de la imagen.
    def crearArregloMB16x16(self, pixelMatrix, w, h):
        arrMB = []
        for x in range (0,w, 16):
            for y in range (0, h, 16):
                pixelesMB = []
                for i in range (x,x+15):
                    for j in range (y,y+15):
                        pixelesMB.append(pixelMatrix[i][j])
                MB = MacroBloque(pixelesMB, x, y)
                arrMB.append(MB) 
        return arrMB

    def crearArregloMBpxp(self, pixelMatrix, w, h):
        arrMB = []
        for x in range (0,w-15):
            for y in range (0, h-15):
                pixelesMB = []
                for i in range (x,x+15):
                    for j in range (y,y+15):
                        pixelesMB.append(pixelMatrix[i][j])
                MB = MacroBloque(pixelesMB, x, y)
                arrMB.append(MB) 
        return arrMB

    #Lee y estructura la informacion de las imagenes en Macrobloques.
    def leer(self):
        im1 = Image.open(self.frame1)
        im2 = Image.open(self.frame2)

        w = im1.width
        h = im1.height

        if (im1.mode == 'RGB' and im2.mode == 'RGB'):
            im1.quantize()
            im2.quantize()
        
        matrizPixel1 = self.llenarMatrizPixel(im1)
        matrizPixel2 = self.llenarMatrizPixel(im2)

        ArrMB1 = self.crearArregloMB16x16(matrizPixel1, w, h)
        print(len(ArrMB1))
        ArrMB2 = self.crearArregloMBpxp(matrizPixel2, w, h)
        print(len(ArrMB2))
        datos = (ArrMB1, ArrMB2, w, h)
        return datos
        

File: SecuencialC++/lector.cpp
#pragma once
#include <iostream>
#include <sstream>
#include <vector>
#include "macrobloque.cpp"
#include <string>

using namespace std;

class lector
{
  private:
    char* frame1;
    char* frame2;
    int w;
    int h;

  public:
    lector(){

    }

    lector(char* frame1, char* frame2)
    {
        this->frame1 = frame1;
        this->frame2 = frame2;
    }

    vector<vector<string> > llenarMatrizPixeles(char *filename)
    {
        int c;
        FILE *f = fopen(filename, "rb");
        unsigned char info[54];
        fread(info, sizeof(unsigned char), 54, f); // read the 54-byte header

        int width = *(int *)&info[18];
        int height = *(int *)&info[22];

        this->w = width;
        this->h = height;

        int size = 3 * width * height;
        unsigned char *data = new unsigned char[size]; // allocate 3 bytes per pixel
        fread(data, sizeof(unsigned char), size, f);   // read the rest of the data at once
        fclose(f);

        vector<string> pixeles = vector<string>(width * height);

        for (c = 0; c < size; c += 3)
        {
            stringstream ss;
            ss << data[c];
            ss<< data[c + 1];
            ss<< data[c + 2];
            string p;
            ss >> p;
            pixeles.push_back(p);
        }

        vector<vector<string> > matrizPixeles = vector<vector<string> >(width);
        for (int i = 0; i < matrizPixeles.size(); ++i)
        {
            vector<string> columna = vector<string>(height);
            matrizPixeles.push_back(columna);
            for (int j = 0; j < matrizPixeles[i].size(); ++j)
            {
                matrizPixeles[i].push_back(pixeles.front());
                pixeles.erase(pixeles.begin());
            }
        }
        return matrizPixeles;
    }

    vector<macrobloque> crearArregloMB16x16(vector<vector<string> > pixelMatrix, int w, int h){
        vector<macrobloque> arrMB;
        for (int x = 0; x < w; x + 16){
            for (int y = 0; y < h; y + 16){
                vector<string> pixelesMB;
                for (int i = x; i < x + 15; ++i){
                    for (int j = y; j < j + 15; ++j){
                        pixelesMB.push_back(pixelMatrix[i][j]);
                    }
                }
                macrobloque MB = macrobloque(pixelesMB, x, y);
                arrMB.push_back(MB);
            }
        }
        return arrMB;
    }

    vector<macrobloque> crearArregloMBpxp(vector<vector<string> > pixelMatrix, int w, int h){
        vector<macrobloque> arrMB;
        for (int x = 0; x < w-15; ++x){
            for (int y = 0; y<h-15; ++y){
                vector<string> pixelesMB;
                for (int i = x; i < x + 15; ++i){
                    for (int j = y; j < j + 15; ++j){
                        pixelesMB.push_back(pixelMatrix[i][j]);
                    }
                }
                macrobloque MB = macrobloque(pixelesMB, x, y);
                arrMB.push_back(MB);
            }
        }
        return arrMB;
    }

    vector<vector<macrobloque> > leer(){
        vector<vector<string> > matrizPixel1 = llenarMatrizPixeles(this->frame1);
        vector<vector<string> > matrizPixel2 = llenarMatrizPixeles(this->frame2);

        vector<macrobloque> ArrMB1 = crearArregloMB16x16(matrizPixel1, w, h);
        vector<macrobloque> ArrMB2 = crearArregloMBpxp(matrizPixel2, w, h);

        vector<vector<macrobloque> > datos;

        datos.push_back(ArrMB1);
        datos.push_back(ArrMB2);

        return datos;
    }
};